### PR TITLE
ANN: Check E0116 "Inherent impls should be in same crate"

### PR DESCRIPTION
**Pull request by dima74**
Related: #6308, #6777, #886

changelog: Detect [E0116](https://doc.rust-lang.org/error-index.html#E0116) compiler error

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt [hunk 1]
```diff
@@ -4009,4 +4009,17 @@ class RsErrorAnnotatorTest : RsAnnotatorTestBase(RsErrorAnnotator::class) {
...
     """)
+
+    @MockEdition(CargoWorkspace.Edition.EDITION_2018)
+    fun `test E0116 inherent impls should be in same crate`() = checkByFileTree("""
+    //- lib.rs
+        pub struct ForeignStruct {}
+    //- main.rs
+        /*caret*/
+        use test_package::ForeignStruct;
+        impl <error descr="Cannot define inherent `impl` for a type outside of the crate where the type is defined [E0116]">ForeignStruct</error> {}
```
Comment (ortem) [inline src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt:4020]: What about error annotations for aliased types? E.g.
```Rust
type Foo = ForeignStruct;
impl Foo {}
```

Comment (dima74) [reply src/test/kotlin/org/rust/ide/annotator/RsErrorAnnotatorTest.kt]: Thanks, updated tests

**Review by ortem — APPROVED** [review]
LGTM
bors r+

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -811,6 +812,16 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
         RsDiagnostic.ImplBothCopyAndDropError(element).addToHolder(holder)
     }
 
+    // E0116: Cannot define inherent `impl` for a type outside of the crate where the type is defined
+    private fun checkInherentImplSameCrate(holder: RsAnnotationHolder, impl: RsImplItem) {
+        if (impl.traitRef != null) return  // checked in [checkTraitImplOrphanRules]
+        val typeReference = impl.typeReference ?: return
+        val struct = (typeReference.type as? TyAdt)?.item ?: return
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:819]: What's about other types?
For example, 
```rust
impl dyn Default {}
```

also should produce E0116

**Pull request merged**